[GStreamer][WebAudio][Realtek] Fix WebAudio support on realtek platform

calvaris opened the pull request:
#### b9dec87d19368cb37dfd62fbab8d8b2ac4d94065
<pre>
[GStreamer][WebAudio][Realtek] Fix WebAudio support on realtek platform
<a href="https://bugs.webkit.org/show_bug.cgi?id=253434">https://bugs.webkit.org/show_bug.cgi?id=253434</a>

Reviewed by Philippe Normand.

This commit adds some custom settings for GstRTKAudioSink and excludes
some hardware decoders from participating in WebAudio processing.

Based on a patch by Pawel Lampe &lt;pawel.lampe@gmail.com&gt;.

* Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.cpp:
(WebCore::AudioDestinationGStreamer::AudioDestinationGStreamer):
* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::decodebinAutoplugSelectCallback):
(WebCore::AudioFileReader::~AudioFileReader):
(WebCore::AudioFileReader::decodeAudioForBusCreation):
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/261568@main">https://commits.webkit.org/261568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a606f1e185faae328e0b7c3597b2bb9d6c7e29f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112159 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3959 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120814 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116222 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22641 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/12397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/4491 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16825 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100000 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105220 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98782 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/542 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45829 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100572 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13703 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/577 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11826 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14381 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/12397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101909 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19749 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52576 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31864 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8063 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16174 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109952 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27165 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.cpp:
```diff
@@ -138,6 +138,13 @@ AudioDestinationGStreamer::AudioDestinationGStreamer(AudioIOCallback& callback,
+#if PLATFORM(REALTEK)
+        if (!g_strcmp0(G_OBJECT_TYPE_NAME(object), "GstRTKAudioSink")) {
+            g_object_set(object, "media-tunnel", FALSE, nullptr);
+            g_object_set(object, "audio-service", TRUE, nullptr);
```
Please collapse this to a single `g_object_set()`.

philn commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -100,9 +99,28 @@ int decodebinAutoplugSelectCallback(GstElement*, GstPad*, GstCaps*, GstElementFa
+#endif
+    };
+    for (const auto& pluginToSkip : pluginsToSkip) {
+        if (pluginToSkip == StringView::fromLatin1(gst_plugin_feature_get_name(GST_PLUGIN_FEATURE(factory))))
```
Use a var for right hand side instead of creating the same string 10 times?

philn commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -152,9 +170,7 @@ AudioFileReader::~AudioFileReader()
 
     if (m_decodebin) {
         g_signal_handlers_disconnect_matched(m_decodebin.get(), G_SIGNAL_MATCH_DATA, 0, 0, nullptr, nullptr, this);
-#if PLATFORM(BCM_NEXUS) || PLATFORM(BROADCOM)
```
Please add `PLATFORM(REALTEK)` here instead of removing the ifdef

calvaris committed:
[GStreamer][WebAudio][Realtek] Fix WebAudio support on realtek platform
https://bugs.webkit.org/show_bug.cgi?id=253434

Reviewed by Philippe Normand.

This commit adds some custom settings for GstRTKAudioSink and excludes
some hardware decoders from participating in WebAudio processing.

Based on a patch by Pawel Lampe <pawel.lampe@gmail.com>.

* Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.cpp:
(WebCore::AudioDestinationGStreamer::AudioDestinationGStreamer):
* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::decodebinAutoplugSelectCallback):
(WebCore::AudioFileReader::~AudioFileReader):
(WebCore::AudioFileReader::decodeAudioForBusCreation):
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/261568@main
